src/main/java/com/google/edith/services/SearchServiceImpl.java

```diff
@@ -0,0 +1,153 @@
+              // Datastore returns as double even when uploaded as float
+              (float) ((double) entity.getProperty("price")),
+              items.toArray(new Item[items.size()]));
+
+      receipts.add(receipt);
+    }
+    return ImmutableList.copyOf(receipts);
+  }
+
+  @Override
+  public ImmutableList<Item> createItemObjects(ImmutableList<Entity> entities) {
+    List<Item> itemsList = new ArrayList<>();
+
+    for (Entity entity : entities) {
+      Item item =
+          Item.builder()
+              .setUserId((String) entity.getProperty("userId"))
+              .setName((String) entity.getProperty("name"))
+              .setPrice((double) entity.getProperty("price"))
+              .setQuantity((long) entity.getProperty("quantity"))
+              .setDate((String) entity.getProperty("date"))
+              .setCategory((String) entity.getProperty("category"))
+              .setExpiration((String) entity.getProperty("expireDate"))
+              .build();
+
+      itemsList.add(item);
+    }
+    return ImmutableList.copyOf(itemsList);
+  }
+
+  @Override
+  public ImmutableList<Entity> findEntityFromDatastore(
+      String name, String date, String kind, String sortOrder, String sortOnProperty) {
+    Query query = prepareQuery(name, date, kind, sortOrder, sortOnProperty);
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> entities = results.asList(FetchOptions.Builder.withLimit(Integer.MAX_VALUE));
+    return ImmutableList.copyOf(entities);
+  }
+
+  /**
+   * Prepares a query for the search on datastore.
+   *
+   * @param name -name property of the entity.
+   * @param date - date property of the entity.
+   * @param kind - kind of the entity stored in datastore.
+   * @param sortOrder - order to sort the entities.
+   * @param sortOnProperty - property on which to sort the order.
+   * @return Query - query to be made on the datastore
+   */
+  private Query prepareQuery(
+      String name, String date, String kind, String sortOrder, String sortOnProperty) {
+    Query query;
+    Filter entityFilter = prepareFilter(name, date);
+
+    if (kind.equals("Receipt")) {
+      query = new Query("Receipt").setFilter(entityFilter);
+    } else {
+      query = new Query("Item").setFilter(entityFilter);
+    }
+
+    if (sortOrder.equals("Ascending"))
+      query = query.addSort(sortOnProperty, SortDirection.ASCENDING);
+    if (sortOrder.equals("Descending"))
+      query = query.addSort(sortOnProperty, SortDirection.DESCENDING);
+    return query;
+  }
+
+  /**
+   * Creates a filter either on name and date.
+   *
+   * @param name -name property of the entity.
+   * @param date - date property of the entity.
+   * @return Filter -filter options to filer on entites.
+   */
+  private Filter prepareFilter(String name, String date) {
+    String loggedInUserId = userService.getCurrentUser().getUserId();
+    List<Filter> filters = new ArrayList<>();
```

Comment:
I'm not sure if it's necessary, but you could use an ImmutableList here too, and just put

Reply:
I am building a filter here with each filter obtained from the user. I am not quite sure it we can make ImmutableList and add to the list as we go ? lmk if I am missing something.